[RFC] Add VectorSpaceVector class.

Rombur opened the pull request:
This is another step to add the new vector interface #745 This is the interface that all the vector classes will have to implement. Please let me know, if I should add /remove some functions.

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Copy constructor.
+     */
+    virtual VectorSpaceVector(const VectorSpaceVector &V) = 0;
```
I'm confused -- can you even have virtual constructors? And could they be abstract (=0)?

Rombur replied:
You can't, my bad.

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+     * Imports all the elements present in the vector's IndexSet from the input
+     * vector @p read_write_vector.
+     */
+    virtual VectorSpaceVector<Number> operator= (const ReadWriteVector<Number> &read_write_vector) = 0;
```
`operator=` should return a reference.

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+     * Multiply the entire vector by a fixed factor.
+     */
+    template <typename Number2>
+    virtual VectorSpaceVector<Number> operator*= (const Number2 factor) = 0;
```
You can't have templated virtual functions.

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+     * Divide the entire vector by a fixed factor.
+     */
+    template <typename Number2>
+    virtual VectorSpaceVector<Number> operator/= (const Number2 factor) = 0;
```
same here and likely below

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+
+    /**
+     * Returns the global size of the vector, equal to the sum of the number of
+     * locally owned indices amon all processors.
```
amon -> among

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+     * only swaps the pointers to the data of the two vectors and therefore
+     * does not need to allocate temporary storage and move data around.
+     */
+    virtual void swap(VectorSpaceVector<Number> &V) = 0;
```
I don't think you need this -- swapping objects only makes sense if you actually know the concrete data types of both objects.

bangerth commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,219 @@
+     * Return a reference to the MPI communicator object in use with this
+     * vector.
+     */
+    virtual const MPI_Comm &get_mpi_communicator() const = 0;
```
This only makes sense for certain types of implementations of this interface. I'd remove this function and only put it into those classes where it is possible.

Rombur replied:
This should always work, even without MPI because we have a typedef on it. But I can remove it.

bangerth replied:
I know. But it's a function that not all derived classes can reasonably implement. If it turns out that we need this in some sort of generic way, then we need to have an intermediate class `ParallelVectorSpaceVector` that adds this function. But we'll get there when we get there.

bangerth commented:
I would appreciate if someone else could also take a look.

@Rombur -- could you also squash the two commits, as they seem to be working against each other :-)

bangerth commented:
Would someone else mind taking a look?

kronbichler commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,181 @@
+  {
+  public:
+    typedef types::global_dof_index                           size_type;
+    typedef typename numbers::NumberTraits<Number>::real_type real_type;
```
There should probably also be a typedef to value_type in here; I guess there is no need for the full set of iterators/pointers because those are implementation-specific and do not make sense in a base class.

tjhei commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,181 @@
+    /**
+     * Multiple addition of scaled vectors, i.e. <tt>*this += a*V+b*W</tt>.
+     */
+    virtual void add(const Number a, const VectorSpaceVector<Number> &VW,
```
typo, should be V not VW. :-)

tjhei commented on include/deal.II/lac/vector_space_vector.h:
```diff
@@ -0,0 +1,181 @@
+     * Imports all the elements present in the vector's IndexSet from the input
+     * vector @p read_write_vector.
+     */
+    virtual VectorSpaceVector<Number> &operator= (const ReadWriteVector<Number> &read_write_vector) = 0;
```
This is an "import" operation and assumes that ```this``` has the correct shape to import, right? Maybe we should avoid ```operator=``` for this and call it ```import```?

kronbichler replied:
Is this the only way to go from a read-write vector to a vector space vector?

I am thinking about the general workflow in a time-dependent problem and how I should use read-write versus vector-space vectors. This interface means that I will mostly have vector-space vectors to represent all my solution vectors. This way, I can add vectors by add/sadd/equ and do the usual stuff on vectors. When I go to an assembly routine which involves old time step data, I will copy the data to a readwrite vector and do whatever is necessary. I will have another read-write vector to collect my residual. In the end of assembly, I will call operator= on a vector-space vector to gather all data. So doing the time steps look fine. If efficiency is important, one might not actually copy the data but rather re-use some pointers in a suitable way.

However, what do we do with the function interpolating initial conditions? For that, I need to access individual vector elements, i.e., a ReadWriteVector. However, I need to define a different 'combine mode' here than for the assembly ('insert' rather than 'add') because otherwise I will duplicate some vector entries. (Or is the DoFHandler supposed to only insert locally owned data? That calls for errors.)

Thus, I see the need for a VectorOperation argument here.

bangerth replied:
I think we don't need this function here at all. The current class has no concrete members, it only provides an abstract interface. There is nothing to copy here. The copying has to happen in derived classes.

@kronbichler -- yes, we will need some kind of "combine" mode argument in assignments.

tjhei commented:
I think it make sense to add a function that returns the ```MPI_Comm``` used by the object.

kronbichler commented:
Apart from those comments, I think that the basic interface looks fine.

tjhei commented:
Agreed, interface looks fine. Other things we should probably define somewhere in the documentation:
- constructors (MPI_COMM and IndexSet, copy constructor, ... ?)
- reinit() calls

other operations worth thinking about:
- min(), max(), 
- operator* for scalar product
- swap

kronbichler commented:
Indeed, operator* is missing and is definitely required (`add_and_dot` is a specialization/optimization for some context but it certainly does not cover them all).

bangerth commented:
@tjhei -- @Rombur actually had a function returning an MPI_Comm originally and I talked him out of it. The point is that not all classes will be able to implement it. For example, completely local vectors or vectors that live on GPUs won't have it. If we want, we can build a separate class on top of this one that adds this one interface. Let's see whether we need it.

For `swap` you need to know the two concrete types. If you have two objects of *different* derived classes, then you can't reasonably expect `swap` to work, so having an interface that only takes references to base classes doesn't make any sense.

Regarding constructors and reinit functions: These can't be made virtual, but you are right that we should document somewhere which variants we want. Can we leave adding this to a time when we actually implement derived classes?

@Rombur -- can you add operator* (the scalar product) and `add_and_dot`?

Rombur commented:
I have removed ```operator=```, added ```operator*``` (```add_and_dot``` was already there), fix the typo, and added ```value_type```. 

@tjhei I thought about ```min``` and ```max``` but for now I'd like the interface to be as small as possible. If we need these functions, we can always add them later.

tjhei commented:
@bangerth Of course constructors/reinit/swap won't be virtual functions in the base class, but I think we should require that all classes implement them. So that is why I was suggesting "define somewhere in the documentation".

I agree that min/max is not important.

Rombur commented:
If we agree on this interface for now, can we merge this. I want to implement the first vector that we will derive from ```ReadWriteVector``` and ```VectorSpaceVector```. This should help defining the interface in more details.

Rombur commented:
Can we merge this please?

tjhei commented:
Sure, we can always edit things later on.
